Dialer.RequireStartTLS to configure STARTTLS security

sfllaw opened the pull request:
It is possible to use a man-in-the-middle attack to strip STARTTLS security from an SMTP connection. Because all modern SMTP servers support STARTTLS, we recommend that `Dialer.RequireStartTLS` be set to `mail.MandatoryStartTLS`.

For backwards compatibility, this library defaults to `mail.OpportunisticStartTLS`, which means that the client will fallback to a cleartext connection if the SMTP server does not support the STARTTLS extension.

This has no effect on SMTPS connections where TLS wraps the connection itself.

Closes go-mail/mail#23.

sfllaw committed:
Fix typo

sfllaw committed:
Extract mockClient from doTestSendMail

In order to prevent `mail.doTestSendMail` from being overwhelmed by
parameters, stop creating `mockClient` inside that method and have its
caller create it.

Since there are `testSendMail` and `testSendMailTimeout` helpers, they
create the `mockClient`.

sfllaw committed:
mockClient.startTLS support

When `mockClient.startTLS` is true, then an SMTP transaction asking
for the STARTTLS extension will find it available. Otherwise, it is
missing.

sfllaw committed:
Dialer.RequireStartTLS to configure STARTTLS security

It is possible to use a man-in-the-middle attack to strip STARTTLS
security from an SMTP connection. Because all modern SMTP servers
support STARTTLS, we recommend that `Dialer.RequireStartTLS` be set to
`mail.MandatoryStartTLS`.

For backwards compatibility, this library defaults to
`mail.OpportunisticStartTLS`, which means that the client will
fallback to a cleartext connection if the SMTP server does not support
the STARTTLS extension.

This has no effect on SMTPS connections where TLS wraps the connection
itself.

sfllaw commented:
@ivy Please review this draft patch. I'd be happy to take any suggestions on how to make things simpler when it comes to setting options on the `Dialer`.

ivy commented on smtp.go:
```diff
@@ -27,9 +27,17 @@ type Dialer struct {
 	// extension instead.
 	SSL bool
-	// TSLConfig represents the TLS configuration used for the TLS (when the
+	// TLSConfig represents the TLS configuration used for the TLS (when the
```
Good catch!

ivy commented on smtp.go:
```diff
@@ -27,9 +27,17 @@ type Dialer struct {
+	// but we recommend MandatoryStartTLS for all modern SMTP servers.
+	//
+	// This option has no effect if SSL is set to true.
+	RequireStartTLS StartTLSSecurity
```
I think `StartTLSPolicy` would be a more fitting name for both `RequireStartTLS` and `StartTLSSecurity`.

sfllaw replied:
Fixed in 0c066ca.

ivy commented on smtp.go:
```diff
@@ -95,8 +103,16 @@ func (d *Dialer) Dial() (SendCloser, error) {
+	if !d.SSL && d.RequireStartTLS != NoStartTLS {
+		ok, _ := c.Extension("STARTTLS")
+		if !ok && d.RequireStartTLS == MandatoryStartTLS {
+			err := fmt.Errorf(
```
We should export an error for this (e.g. `ErrStartTLSUnsupported`).

sfllaw replied:
Fixed in d3eed43, but named `StartTLSUnsupportedError` to follow Go conventions: https://blog.golang.org/error-handling-and-go

ivy replied:
That much better than my suggestion. I like your use of `StartTLSPolicy.String()` for handling the error message.😄

ivy reviewed (changes requested):
I only have a couple nitpicks. Everything else looks good to me!

sfllaw committed:
Rename RequireStartTLS and StartTLSSecurity to StartTLSPolicy

sfllaw committed:
Dial returns StartTLSUnsupportedError on failure

Instead of returning a bare error, `Dial` returns
`StartTLSUnsupportedError` when `Dialer.StartTLSPolicy` requires
STARTTLS, but it is unsupported by the server.

sfllaw committed:
Merge branch 'v2' into Dialer.RequireStartTLS

ivy commented:
@sfllaw Thanks for implementing this along with the changes I suggested. :+1:

sfllaw commented:
@ivy Thanks for catching the missing change in `CHANGELOG.md`.

Do you have plans to release 2.2.0 soon? I'd like to tighten up STARTTLS in some of our code. :smile:

ivy commented:
@sfllaw Sure thing, I tagged and released 2.2.0. Thanks again for helping tighten up security! :+1:
